src/service/storage-impl.js

```diff
@@ -0,0 +1,391 @@
+   * Returns the promise that yields the value of the property for the specified
+   * key.
+   * @param {string} name
+   * @return {!Promise<*>}
+   * @override
+   */
+  get(name) {
+    return this.getStore_().then(store => store.get(name));
+  }
+
+  /**
+   * Saves the value of the specified property. Returns the promise that's
+   * resolved when the operation completes.
+   * @param {string} name
+   * @param {*} value
+   * @return {!Promise}
+   * @override
+   */
+  set(name, value) {
+    assert(typeof value == 'boolean', 'Only boolean values accepted');
+    return this.saveStore_(store => store.set(name, value));
+  }
+
+  /**
+   * Removes the specified property. Returns the promise that's resolved when
+   * the operation completes.
+   * @param {string} name
+   * @return {!Promise}
+   * @override
+   */
+  remove(name) {
+    return this.saveStore_(store => store.remove(name));
+  }
+
+  /**
+   * @return {!Promise<!Store>}
+   * @private
+   */
+  getStore_() {
+    if (!this.storePromise_) {
+      this.storePromise_ = this.whenStarted_
+          .then(() => this.binding_.loadBlob(this.origin_))
+          .then(blob => blob ? JSON.parse(atob(blob)) : {})
+          .catch(reason => {
+            log.error(TAG, 'Failed to load store: ', reason);
+            return {};
+          })
+          .then(obj => new Store(obj));
+    }
+    return this.storePromise_;
+  }
+
+  /**
+   * @param {function(!Store)} mutator
+   * @return {!Promise}
+   * @private
+   */
+  saveStore_(mutator) {
+    return this.getStore_()
+        .then(store => {
+          mutator(store);
+          const blob = btoa(JSON.stringify(store.obj));
+          return this.binding_.saveBlob(this.origin_, blob);
+        })
+        .then(this.broadcastReset_.bind(this));
+  }
+
+  /** @private */
+  listenToBroadcasts_() {
+    this.viewer_.onBroadcast(message => {
+      if (message['type'] == 'amp-storage-reset' &&
+              message['origin'] == this.origin_) {
+        log.fine(TAG, 'Received reset message');
+        this.storePromise_ = null;
+      }
+    });
+  }
+
+  /** @private */
+  broadcastReset_() {
+    log.fine(TAG, 'Broadcasted reset message');
+    this.viewer_.broadcast({
+      'type': 'amp-storage-reset',
+      'origin': this.origin_
+    });
+  }
+}
+
+
+/**
+ * The implementation of store logic for get, set and remove.
+ *
+ * The structure of the store is equivalent to the following typedef:
+ * ```
+ * {
+ *   vv: !Array<!{
+ *     k: string,
+ *     v: *,
+ *     t: time
+ *   }>
+ * }
+ * ```
+ *
+ * @private Visible for testing only.
+ */
+export class Store {
+  /**
+   * @param {!JSONObject} obj
+   * @param {number=} opt_maxValues
+   */
+  constructor(obj, opt_maxValues) {
+    /** @const {!JSONObject} */
+    this.obj = obj;
+
+    /** @private @const {number} */
+    this.maxValues_ = opt_maxValues || MAX_VALUES_PER_ORIGIN;
+
+    /** @private @const {!Array<!JSONObject>} */
+    this.values_ = obj['vv'] || [];
```

Comment:
Possible of course. The set is very small - forced to be under 8 elements. It simplifies search and removal of old values. And it also seemed a bit better as a format. But I can definitely switch to a map if you feel that format would improve.